File: dynamic-programming/bottom-up/stock-trading/MultiStockTrader.cpp
class MultiStockTrader {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0) {
            return 0;
        }
        int soldLast = 0;
        int holdLast = INT_MIN;
        for (int curPrice : prices) {
            int soldPrev = soldLast;
            soldLast = max(soldLast, holdLast + curPrice);
            holdLast = max(holdLast, soldPrev - curPrice);
        }
        return soldLast;
    }
};

File: dynamic-programming/bottom-up/stock-trading/MultiStockTrader.java

public class MultiStockTrader {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        int soldLast = 0;
        int holdLast = Integer.MIN_VALUE;
        for (int price : prices) {
            int soldPrev = soldLast;
            soldLast = Math.max(soldLast, holdLast + price);
            holdLast = Math.max(holdLast, soldPrev - price);
        }
        return soldLast;
    }
}
